platform-dependent-components/problem-solver/cxx/generate-response-module/agent/GenerateResponseAgent.cpp

```diff
@@ -0,0 +1,182 @@
+}
+
+void GenerateResponseAgent::validateAddrWithInvalidParamException(ScAddr const & addr, std::string const & paramName)
+{
+  SC_CHECK_PARAM(addr, "parameter '" + paramName + "' is not valid");
+}
+
+void GenerateResponseAgent::finishWorkWithMessage(
+    std::string const & message,
+    ScAddr const & questionNode,
+    bool isSuccess)
+{
+  SC_LOG_ERROR(Constants::generateAnswerAgentClassName + ": " + message);
+  utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, false);
+}
+
+bool GenerateResponseAgent::checkAction(ScAddr const & actionAddr)
+{
+  return m_memoryCtx.HelperCheckEdge(Keynodes::action_generate_response, actionAddr, ScType::EdgeAccessConstPosPerm);
+}
+
+ScAddr GenerateResponseAgent::getMessageClassAddr(ScAddr const & messageAddr)
+{
+  ScIterator3Ptr classIterator =
+      m_memoryCtx.Iterator3(ScType::NodeConstClass, ScType::EdgeAccessConstPosPerm, messageAddr);
+
+  while (classIterator->Next())
+  {
+    ScAddr const & messageClassAddr = classIterator->Get(0);
+    bool isMessageClass = m_memoryCtx.HelperCheckEdge(
+        Keynodes::concept_intent_possible_class, messageClassAddr, ScType::EdgeAccessConstPosPerm);
+    if (isMessageClass)
+      return messageClassAddr;
+  }
+
+  return {};
+}
+
+bool GenerateResponseAgent::attachAnswer(
+    ScAddr const & messageAnswer,
+    ScAddr const & messageAddr,
+    ScAddr const & answerAddr)
+{
+  if (!messageAnswer.IsValid())
+    return false;
+
+  ScAddr const & edgeBetweenMessageAndMessageAnswer =
+      m_memoryCtx.CreateEdge(ScType::EdgeDCommonConst, messageAddr, messageAnswer);
+  m_memoryCtx.CreateEdge(ScType::EdgeAccessConstPosPerm, Keynodes::nrel_reply, edgeBetweenMessageAndMessageAnswer);
+
+  m_memoryCtx.CreateEdge(ScType::EdgeAccessConstPosTemp, answerAddr, messageAnswer);
+
+  return true;
+}
+
+ScAddrVector GenerateResponseAgent::getMessageParameters(ScAddr const & messageAddr)
+{
+  static ScTemplate templ;
+  templ.Clear();
+  templ.Triple(
+      Keynodes::concept_entity_possible_class,
+      ScType::EdgeAccessVarPosPerm,
+      ScType::NodeVarClass >> "_message_entity_class");
+  templ.Quintuple(
+      "_message_entity_class",
+      ScType::EdgeDCommonVar,
+      ScType::NodeVarRole >> Constants::roleRelationVarName,
+      ScType::EdgeAccessVarPosPerm,
+      Keynodes::nrel_entity_possible_role);
+  templ.Quintuple(
+      messageAddr,
+      ScType::EdgeAccessVarPosPerm,
+      ScType::NodeVar >> Constants::messageParamVarName,
+      ScType::EdgeAccessVarPosPerm,
+      Constants::roleRelationVarName);
+
+  ScTemplateSearchResult result;
+  m_memoryCtx.HelperSearchTemplate(templ, result);
+
+  if (result.IsEmpty())
+  {
+    SC_LOG_WARNING(Constants::generateAnswerAgentClassName + ": no params founded");
+    return {};
+  }
+
+  ScAddrVector parameters;
+
+  ScAddr const & messageParameter = result[0][Constants::messageParamVarName];
+  ScAddr const & roleRelation = result[0][Constants::roleRelationVarName];
```

Comment:
if we need only first result, then possibly it's worth to use smart search

Reply:
We'll probably use more than one parameter in the future. Then we'll just wrap it in a loop